This exercise will require you to pull some data from the Qunadl API. Qaundl is currently the most widely used aggregator of financial market data.

As a first step, you will need to register a free account on the http://www.quandl.com website.

After you register, you will be provided with a unique API key, that you should store:

In [1]:
# Store the API key as a string - according to PEP8, constants are always named in all upper case
API_KEY = '7up7a3-xNp8abz4VJGsq'

Qaundl has a large number of data sources, but, unfortunately, most of them require a Premium subscription. Still, there are also a good number of free datasets.

For this mini project, we will focus on equities data from the Frankfurt Stock Exhange (FSE), which is available for free. We'll try and analyze the stock prices of a company called Carl Zeiss Meditec, which manufactures tools for eye examinations, as well as medical lasers for laser eye surgery: https://www.zeiss.com/meditec/int/home.html. The company is listed under the stock ticker AFX_X.

You can find the detailed Quandl API instructions here: https://docs.quandl.com/docs/time-series

While there is a dedicated Python package for connecting to the Quandl API, we would prefer that you use the *requests* package, which can be easily downloaded using *pip* or *conda*. You can find the documentation for the package here: http://docs.python-requests.org/en/master/ 

Finally, apart from the *requests* package, you are encouraged to not use any third party Python packages, such as *pandas*, and instead focus on what's available in the Python Standard Library (the *collections* module might come in handy: https://pymotw.com/3/collections/).
Also, since you won't have access to DataFrames, you are encouraged to us Python's native data structures - preferably dictionaries, though some questions can also be answered using lists.
You can read more on these data structures here: https://docs.python.org/3/tutorial/datastructures.html

Keep in mind that the JSON responses you will be getting from the API map almost one-to-one to Python's dictionaries. Unfortunately, they can be very nested, so make sure you read up on indexing dictionaries in the documentation provided above.

In [2]:
# First, import the relevant modules
import requests
import json

In [3]:
# Now, call the Quandl API and pull out a small sample of the data (only 5 days) to get a glimpse
# into the JSON structure that will be return
r=requests.get('https://www.quandl.com/api/v3/datasets/FSE/AFX_X.json?api_key=7up7a3-xNp8abz4VJGsq&start_date=2018-07-22')

In [4]:
# Inspect the JSON structure of the object you created, and take note of how nested it is,
# as well as the overall structure
r.json()['dataset']

{'id': 10095370,
 'dataset_code': 'AFX_X',
 'database_code': 'FSE',
 'name': 'Carl Zeiss Meditec (AFX_X)',
 'description': 'Stock Prices for Carl Zeiss Meditec (AFX) from the Frankfurt Stock Exchange.<br><br>Trading System: Xetra<br><br>ISIN: DE0005313704',
 'refreshed_at': '2018-10-01T22:38:27.471Z',
 'newest_available_date': '2018-10-01',
 'oldest_available_date': '2000-06-07',
 'column_names': ['Date',
  'Open',
  'High',
  'Low',
  'Close',
  'Change',
  'Traded Volume',
  'Turnover',
  'Last Price of the Day',
  'Daily Traded Units',
  'Daily Turnover'],
 'frequency': 'daily',
 'type': 'Time Series',
 'premium': False,
 'limit': None,
 'transform': None,
 'column_index': None,
 'start_date': '2018-07-22',
 'end_date': '2018-10-01',
 'data': [['2018-10-01',
   None,
   73.55,
   72.1,
   72.95,
   None,
   214973.0,
   15638105.0,
   None,
   None,
   None],
  ['2018-09-28',
   None,
   73.55,
   71.8,
   72.5,
   None,
   160513.0,
   11660089.0,
   None,
   None,
   None],
  ['20

These are your tasks for this mini project:

1. Collect data from the Franfurt Stock Exchange, for the ticker AFX_X, for the whole year 2017 (keep in mind that the date format is YYYY-MM-DD).
2. Convert the returned JSON object into a Python dictionary.
3. Calculate what the highest and lowest opening prices were for the stock in this period.
4. What was the largest change in any one day (based on High and Low price)?
5. What was the largest change between any two days (based on Closing Price)?
6. What was the average daily trading volume during this year?
7. (Optional) What was the median trading volume during this year. (Note: you may need to implement your own function for calculating the median.)

1. Collect data from the Frankfurt Stock Exchange, for the ticker AFX_X, for the whole year 2017.

In [52]:
year=requests.get('https://www.quandl.com/api/v3/datasets/FSE/AFX_X.json?api_key=7up7a3-xNp8abz4VJGsq&start_date=2017-01-01&end_date=2018-01-01')

2. Convert the returned JSON object into a Python dictionary

In [53]:
data=year.json()['dataset']

3. Calculate what the highest and lowest opening prices were for the stock in this period

In [59]:
year_data=data['data']
col_names=data['column_names']
print(col_names)
len(year_data)

['Date', 'Open', 'High', 'Low', 'Close', 'Change', 'Traded Volume', 'Turnover', 'Last Price of the Day', 'Daily Traded Units', 'Daily Turnover']


255

In [56]:
import pandas as pd
df=pd.DataFrame(year_data, columns=col_names)
df.head()

Date   Open   High    Low  Close  Change  Traded Volume   Turnover  \
0  2017-12-29  51.76  51.94  51.45  51.76     NaN        34640.0  1792304.0   
1  2017-12-28  51.65  51.82  51.43  51.60     NaN        40660.0  2099024.0   
2  2017-12-27  51.45  51.89  50.76  51.82     NaN        57452.0  2957018.0   
3  2017-12-22  51.05  51.50  50.92  51.32     NaN        71165.0  3641949.0   
4  2017-12-21  51.16  51.52  50.90  51.40     NaN       120649.0  6179433.0   

  Last Price of the Day Daily Traded Units Daily Turnover  
0                  None               None           None  
1                  None               None           None  
2                  None               None           None  
3                  None               None           None  
4                  None               None           None

In [57]:
highest_open=df['Open'].max()
lowest_open=df['Open'].min()
print(highest_open)
print(lowest_open)

53.11
34.0


The highest opening is 53.11 and the lowest opening is 34.0 for this period.

4. What was the largest change in one day?

In [58]:
largest_change=(df['High']-df['Low']).max()
print(largest_change)

2.8100000000000023


The largest change in one day is 2.81.

5. What was the largest change in any two days?

In [64]:
changes_two = [df['Close'][i+1]-df['Close'][i] for i in range(len(year_data)-1)]
from numpy import array
largest_two_day_change= array(changes_two).max()
print(largest_two_day_change)

2.559999999999995


The largest change in any two days is 2.559999

6. What was the average trading?

In [41]:
average_trading=df['Traded Volume'].sum()/len(data)
print(average_trading)

1082224.0952380951


The average trading is 1082224.095

Optional 7. What was the median trading

In [36]:
import numpy as np
median_trading=np.median(df['Traded Volume'])
print(median_trading)

76286.0


The median trading is 76286